astropy/io/fits/column.py

```diff
@@ -1521,6 +1523,11 @@ def _update_column_attribute_changed(self, column, attr, old_value,
             if col is column:
                 break
 
+        if attr == 'name':
+            del self.names
+        elif attr == 'format':
+            del self.formats
```

Comment:
> When you say FITS_rec class to be less reliant on the ColDefs, I wonder for example why `FITS_rec.names tries first to use _coldefs.names and then dtype.names, and not the opposite ?

Exactly--I think the opposite makes more sense.  The way it works currently may be a remnant of when the dtype fields and the "logical" columns represented by the `ColDefs` class were *not* guaranteed to be the same.  The reason for this is the unfortunate existence in FITS of "zero width" columns which must be supported--it's rare but does occur in the wild.  It used to be that the Numpy dtype could not accommodate that, so there were a lot of hacks to make all the columns (including zero-width ones) represented to the user, while only representing the non-zero columns in the actual dtype.

I've gotten a bunch of patches into Numpy, however, to allow fields with zero size to work in a consistent manner, so most of those hacks have been removed (there is still one remnant of this issue: https://github.com/numpy/numpy/pull/6430  Though that reminds me, the fix for that will be in Numpy 1.12 so *in principle* it should be possible to avoid use of the `realign_dtype` hack for Numpy 1.12 and up).

The patch in the gist looks good to me.  I haven't tested it, and can't say with certainty that there are some dark corners lurking where that would actually break.  But I don't *think* there--it's definitely what I had in mind.